### PR TITLE
add is leader checker before any raft write operation

### PR DESCRIPTION
**Pull request by JacksonYao287**
1 add is leader checker before any raft write option
2 change `on_destroy` signature to adapt to homestore

### CODE REVIEW

**Review by xiaoxichen — DISMISSED** [review]
lgtm

File: src/lib/homestore_backend/hs_shard_manager.cpp [hunk 1]
```diff
@@ -172,6 +177,16 @@ ShardManager::AsyncResult< ShardInfo > HSHomeObject::_seal_shard(ShardInfo const
         RELEASE_ASSERT(repl_dev != nullptr, "Repl dev null");
     }
 
+    if (!repl_dev) {
```
[the remark this answers is not in the document]
Comment (JacksonYao287) [reply src/lib/homestore_backend/hs_shard_manager.cpp:180]: I think we can add that. let me do it

[the remark this answers is not in the document]
Comment (JacksonYao287) [reply src/lib/homestore_backend/hs_shard_manager.cpp]: I have removed the duplicated check and refine the commit message, ptal @xiaoxichen

**Comment by xiaoxichen** [comment]
@szmyd  FYI we are adding the status check into HO.    As of now we dont check `is_leader`,  if the IO hit follower,  the local write still happens and then get rejected in `propose_to_raft`,  better to avoid the local write as well. 

Also we hit one corner case in SM long running test,  that IO hit a follower, during local write phase,  HS creates rreq with term T1; leader switch happens between local writes and `propose_to_raft`,  the raft will use T2 which results a conflict.

As you said it is possible the role flip during the IO,  either L-> F or F-> L,    if it is F-> L this request still get reject by the check adding this PR,  if L->F,  it rejected in `propose_to_raft`.

**Review by xiaoxichen — APPROVED** [review]
